### PR TITLE
Add check for alternative union syntax - PEP 604

### LINKED ISSUE #4065: Improve checks for PEP 604 - alternative Union syntax
### Is your feature request related to a problem?
Pylint should be able to detect if PEP 604 (alternative Union Syntax) is used inappropriately.
The PEP will be implemented in **3.10**, so the following should emit error messages in versions **3 - 3.9**:
```py
Alias = str | int  # Should be an error

var: str | int  # Should be an error

def func(arg: int | str):  # Should be an error
    pass

def func2() -> int | str:  # Should be an error
    pass
```

Each line is a `TypeError` in Python, eg:
```
Traceback (most recent call last):
  File "test.py", line 3, in <module>
    Alias = str | int
TypeError: unsupported operand type(s) for |: 'type' and 'type'
```

**With postponed evaluation of annotations enabled (3.7 - 3.9)**
```py
from __future__ import annotations

Alias = str | int  # Should be an error

var: str | int

def func(arg: int | str):
    pass

def func2() -> int | str:
    pass
```
Only the alias definition is evaluated directly and thus an error.

### Current situation
No errors are emitted, even in versions prior to 3.10 and without the future import.

--
https://www.python.org/dev/peps/pep-0604/

### PR DESCRIPTION
**Pull request by cdce8p**
## Steps

- [ ] Add yourself to CONTRIBUTORS if you are a new contributor.
- [x] Add a ChangeLog entry describing what your PR does.
- [ ] If it's a new feature or an important bug fix, add a What's New entry in `doc/whatsnew/<current release.rst>`.
- [x] Write a good description on what the PR does.

## Description
Add check for correct usage of alternative union syntax ([PEP 604](https://www.python.org/dev/peps/pep-0604/)).

**Example Python 3.7 - 3.9**
```py
from __future__ import annotations  # required

var: int | None
def func() -> str | int:
    ...

Alias = int | None  # [unsupported-binary-operation]
```

**Example Python 3.10+**
```py
var: int | None
def func() -> str | int:
    ...

Alias = int | None
```

**Links**
https://www.python.org/dev/peps/pep-0604/
https://mypy.readthedocs.io/en/stable/runtime_troubles.html#using-x-y-syntax-for-unions
https://mypy.readthedocs.io/en/stable/runtime_troubles.html#future-annotations

## Type of Changes

|   | Type |
| ------------- | ------------- |
|   | :bug: Bug fix  |
| ✓  | :sparkles: New feature |
|   | :hammer: Refactoring  |
|   | :scroll: Docs |

## Related Issue

Closes #4065

### CODE REVIEW

**Review by Pierre-Sassoulas — APPROVED** [review]
image


Look like you're going to upgrade from "first time contributor" to "person who implemented all the checks for new typing syntax" as soon as the astroid master is fixed ;)

File: pylint/checkers/typecheck.py [hunk 1]
```diff
@@ -1629,6 +1631,46 @@ def visit_unaryop(self, node):
...
+                n = helpers.object_type(n)
+                if isinstance(n, astroid.ClassDef) and (
+                    n.name == "type"
+                    or any(
+                        [
+                            True
+                            for base in n.bases
+                            if isinstance(base, astroid.Name) and base.name == "type"
+                        ]
+                    )
```
Comment (Pierre-Sassoulas) [inline pylint/checkers/typecheck.py:1667]: Maybe a simplification to do here ?
```suggestion
                    or any(
                        isinstance(base, astroid.Name) and base.name == "type" for base in n.bases
                    )
```

Comment (cdce8p) [reply pylint/checkers/typecheck.py]: Yeah, should have though after that sooner 😅

File: pylint/checkers/typecheck.py [hunk 2]
```diff
@@ -1629,6 +1631,45 @@ def visit_unaryop(self, node):
...
+                    astroid.List,
+                ),
+            )
+        ):
+            for n in (node.left, node.right):
+                n = helpers.object_type(n)
+                if isinstance(n, astroid.ClassDef) and (
+                    n.name == "type"
+                    or any(
+                        [
```
Comment (Pierre-Sassoulas) [inline pylint/checkers/typecheck.py:1662]: We can also remove the list comprehension for better performance (so we do not create the list and just cut short at the first True) :)

Comment (cdce8p) [reply pylint/checkers/typecheck.py]: Changed it

**Comment by cdce8p** [comment]
> image
> 
> Look like you're going to upgrade from "first time contributor" to "person who implemented all the checks for new typing syntax" as soon as the astroid master is fixed ;)

I just would really like to use the new syntax in a project I frequently contribute to that just upgraded to min Python `3.8`. Since we have a lot of beginners, it's just best to have checks in place to catch mistakes early. Especially if it's not exactly clear what works and what doesn't (#4059 😄).

**Comment by cdce8p** [comment]
@Pierre-Sassoulas Can you also tag it with `2.6.1`? I would really like to see it be included in the next release.

**Pull request merged**